keps/sig-scheduling/4328-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,371 @@
+
+The feature can be disabled in Alpha and Beta versions by restarting kube-controller-manager, kube-scheduler and kube-apiserver with the feature-gate off. 
+After the feature-gate is disabled and kube-controller-manager is restarted, 
+`affinityEviction` controller will also be disabled.
+In terms of Stable versions, users can choose to opt-out by not setting the corresponding field.
+
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The `RequiredDuringSchedulingRequiredDuringExecution` node affinity will take effect as expected.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+When this feature is disabled, already running workloads with `RequiredDuringSchedulingRequiredDuringExecution` nodeAffinity will still running and won't be evicted anymore. 
+If the pod with `RequiredDuringSchedulingRequiredDuringExecution` nodeAffinity is updated after the feature is disabled, this field will be silently dropped.
+
+###### What specific metrics should inform a rollback?
+
+We will introduce two new metrics for this feature. Please refer to the "SLIs" section below for their specific meanings.
+If either new metric `pod_deletions_node_affinity_total`  or `pod_deletions_node_affinity_duration_seconds` is larget than expected, users should check their system and may need to rollback.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+N/A
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+- Inspect the nodeAffinity.RequiredDuringSchedulingRequiredDuringExecution configuration.
+- Check the corresponding metrics(details in SLIs part below).
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Events
+  - Event Reason: NodeAffinityEviction
+- [x] API .status
+  - Condition reason: DeletionByAffinityEvictionController 
+- [x] Other (treat as last resort)
+  - Details: Pod with `RequiredDuringSchedulingRequiredDuringExecution` must meet affinity rule during both the scheduling phase and the execution phase.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Metrics
+  - Metric name:
+    - pod_deletions_node_affinity_total(counts the number of Pods deleted because of nodeAffinity since this contoller start)
+    - pod_deletions_node_affinity_duration_seconds(Latency, in seconds, between the time when a node's label has been changed and the pod been deleted via AffinityEvictionController)
```

Comment:
where do you plan to get the time when a node label changed? that's not stored in the API.

Reply:
My thought is we use the time when our controller received a node-label-changed event. So there might be some delay.